### PR TITLE
feat: use ChatGPT backend endpoint (graasp/graasp#641)

### PR DESCRIPTION
**Pull request by ReidyT**
> **Warning**
DO NOT CREATE A RELEASE NOW !

> **Note**
Because some users from Neuchâtel are using `app-code-capsule` with GPT-3 for experimentations, their code capsule **must not** be updated with newer version of ChatGPT. 
For this reason, we have decided to wait before to deploy this new version of the code capsule app. This allows us to avoid writing legacy code that will be deleted anyway when GPT-3 is no longer in use.
Indeed, the other option would have been to manage the ChatBot version in the code capsule (GPT-3, 3.5, 4) allowing the old code capsule to continue to use GPT-3. 
The problem with that is we must set GPT-3 by default to be sure that the users from Neuchâtel are using the same version during their experimentations. This can conduct other users to still using GPT-3, if they don't notice the possibility to change the GPT version. 
So, to avoid adding legacy code and complexity, the PR is in a waiting state. The backend will have the new route to allow other applications to use chatbot if needed.

This PR has some changes:
- Moving the mockServer to the apps-query-client using mirageJS.
- `useChatbotApi` send chat request to the backend with the new message format (graasp/graasp#641).
- `useChatbotApi` is moved to the apps-query-client (#161).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,9 +22,9 @@
     "@fortawesome/free-regular-svg-icons": "6.4.2",
     "@fortawesome/free-solid-svg-icons": "6.4.2",
     "@fortawesome/react-fontawesome": "0.2.0",
-    "@graasp/apps-query-client": "2.0.3",
+    "@graasp/apps-query-client": "github:graasp/graasp-apps-query-client#641-add-a-chatbot-endpoint-for-apps",
```
Comment (spaenleh) [inline package.json:25]: TODO: use version

File: package.json [hunk 2]
```diff
@@ -84,7 +84,9 @@
     "cypress:open": "env-cmd -f ./.env.development cypress open",
     "test": "concurrently -k -s first \"yarn start:test\" \"yarn test:ci\" ",
     "test:ci": "env-cmd -f ./.env.test cypress run --browser chrome --headless && nyc report --reporter=text --reporter=text-summary",
-    "cov:report": "open ./coverage/lcov-report/index.html"
+    "cov:report": "open ./coverage/lcov-report/index.html",
+    "test:chat": "concurrently -k -s first \"yarn start:test\" \"yarn test:cichat\" ",
+    "test:cichat": "env-cmd -f ./.env.test cypress run --spec 'cypress/e2e/codeReview/chatbotInteractions.cy.ts' --browser chrome --headless && nyc report --reporter=text --reporter=text-summary"
```
Comment (spaenleh) [inline package.json:89]: Remove ?

File: src/interfaces/threadMessage.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { APP_DATA_TYPES } from '@/config/appDataTypes';
+
+export interface ThreadMessage {
```
Comment (spaenleh) [inline src/interfaces/threadMessage.ts:3]: Where do we use this, And why does it reference the `APP_DATA_TYPES` ?

File: src/modules/common/CommentThread.tsx [hunk 4]
```diff
@@ -3,6 +3,8 @@ import { useTranslation } from 'react-i18next';
 
 import { CircularProgress, Stack, Typography } from '@mui/material';
 
+import { UserDataType } from '@graasp/apps-query-client';
```
Comment (spaenleh) [inline src/modules/common/CommentThread.tsx:6]: ?

File: src/utils/chatbot.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import { ChatBotMessage, ChatbotRole } from '@graasp/sdk';
+
+import { List } from 'immutable';
+
+import { APP_DATA_TYPES } from '@/config/appDataTypes';
+import { ThreadMessage } from '@/interfaces/threadMessage';
+
+const { buildPostChatBotRoute } = API_ROUTES;
+
+export const chatBotPostUrl = (apiHost: string, itemId: string): string =>
```
Comment (spaenleh) [inline src/utils/chatbot.ts:11]: This route has already been added in the query-client no ?

File: src/interfaces/threadMessage.ts (hunk 3 above)
Comment (ReidyT) [reply src/interfaces/threadMessage.ts:3]: This interface is used in `ChatbotPrompts.tsx` and in `src/utils/chatbot.ts`. It is used to construct the prompt for ChatGPT with the initial prompt, and the history of messages. I updated the `type` from `APP_DATA_TYPES` to a more restrictive values `APP_DATA_TYPES.BOT_COMMENT | APP_DATA_TYPES.COMMENT`. 

The usage of APP_DATA_TYPES values allow the `buildPrompt` to construct the history with defining which message is the assistant and which is the user.

File: src/modules/chatbot/ChatbotPrompts.tsx [hunk 6]
```diff
@@ -99,16 +91,46 @@ const ChatbotPrompts: FC<Props> = ({ line }) => {
...
+
+        const prompt = buildPrompt(
+          currentLinePrompt?.data.initialPrompt,
+          threadMessages,
+          newUserComment,
+        );
+
         postAction({
-          data: { prompt: fullPrompt },
+          data: { prompt: JSON.stringify(prompt) },
```
[the remark this answers is not in the document]
Comment (ReidyT) [reply src/modules/chatbot/ChatbotPrompts.tsx:108]: Yes, it works without stringify, I will remove it.

**Pull request merged**